\begin{frame}[fragile]
  \partpage

\paragraph{Packages required for reproducing the slides}
<<load_packages, cache = FALSE, message=FALSE, warning=FALSE>>=
library(tidyverse)  # opinionated collection of packages for data manipulation
library(FactoMineR) # PCA and oter linear method for dimension reduction
library(factoextra) # fancy plotting for FactoMineR output 
library(NMF)        # Non-Negative Matrix factorisation
library(kernlab)    # Kernel-based methods, among which kernel-PCA
library(MASS)       # Various statistical tools, including metric MDS
library(Rtsne)      # tSNE implementation in R 
library(umap)       # Uniform Manifold Approximation and Projection

theme_set(theme_bw()) # my default theme for ggplot2
@

\end{frame}

\begin{frame}[fragile]
  \frametitle{Companion data set: 'scRNA'}
  \framesubtitle{Subsamples of normalized Single-Cell RNAseq}

\begin{block}{Description: \textcolor{black}{\it subsample of a large data set}}
\small Gene-level expression of 100 representative genes for a collection of 301 cells 
spreaded in 11 cell-lines. Original transcription data are measured by counts obtained by 
\textit{RNAseq} and normalized to be close to a Gaussian distribution.\\

\begin{scriptsize}
\begin{thebibliography}{9}
\bibitem{pollen} Pollen, Alex A., et al. \textcolor{black}{Low-coverage single-cell mRNA sequencing reveals cellular heterogeneity and activated signaling pathways in developing cerebral cortex.} \newblock Nature biotechnology 32.10 (2014): 1053.
\end{thebibliography}
\end{scriptsize}
\end{block}

\begin{figure}
  \includegraphics[width=.9\textwidth]{figures/scRNA-overview}
  \caption{Single Cell RNA sequnencing data: general principle -- {\tiny source: Stephanie Hicks}}
\end{figure}

\end{frame}

\begin{frame}[fragile]
  \frametitle{Companion data set: 'scRNA'}
  \framesubtitle{Brief data summary I}

\paragraph{Data manipulation}

<<scRNA dataset load>>=
load("../../data/scRNA.RData")
scRNA <-  pollen$data %>% t() %>% as_tibble() %>% 
  add_column(cell_type = pollen$celltypes)
@

\paragraph{Cell types}

<<scRNA dataset celltypes, fig.cap='Natural grouping: cell-types'>>=
scRNA %>% dplyr::select(cell_type) %>% summary()  %>% knitr::kable()
@

\end{frame}

\begin{frame}[fragile]
  \frametitle{Companion data set II: 'scRNA'}
  \framesubtitle{Brief data summary II}

\paragraph{Histogram of normalized expression}

<<scRNA expressions>>=
scRNA %>% dplyr::select(-cell_type) %>% pivot_longer(everything()) %>% 
  ggplot() + aes(x = value, fill = name) + geom_histogram(show.legend = FALSE)
@
\end{frame}

\begin{frame}[fragile]
  \frametitle{Companion data set: 'scRNA'}
  \framesubtitle{Principal Component Analysis}

<<scRNA PCA>>=
scRNA %>% PCA(graph = FALSE, quali.sup = which(colnames(scRNA) == "cell_type")) %>% 
  fviz_pca_biplot(select.var = list(contrib = 30), habillage = "cell_type")
@

\end{frame}

\begin{frame}
  \frametitle{PCA (and linear methods) limitations}

  \begin{block}{Do not account for complex pattern}
    \begin{itemize}
      \item Linear methods are powerful for \alert{\bf planar structures}
      \item May fail at describing \alert{\bf manifolds}
    \end{itemize}
  \end{block}
  
  \begin{block}{Fail at preserving local geometry}
    \begin{itemize}
      \item High dimensional data are characterized by \alert{\bf multiscale properties} (local / global structures)
      \item Non Linear projection helps at preserving \alert{\bf local characteristics} of distances
    \end{itemize}
  \end{block}

  \vfill
  
   \begin{figure}
     \includegraphics[scale=0.25]{figures/manifold.pdf}
     \caption{\small Intuition of manifolds and geometry underlying sc-data -- {\tiny source: F. Picard}}
   \end{figure}

\end{frame}

\begin{frame}[fragile]
  \frametitle{Companion data set II: 'mollusk'}
  \framesubtitle{Abundance table (Species counts spread in various sites)}

\begin{block}{Description: \textcolor{black}{\it small size count data}}
\small Abundance of 32 mollusk species in 163 samples. For each sample, 4 additional covariates are known. \\

\begin{scriptsize}
\begin{thebibliography}{9}
\bibitem{mollusk} Richardot-Coulet, M., Chessel D. and Bournaud M. \textcolor{black}{Typological value of the benthos of old beds of a large river. Methodological approach. Archiv f√πr Hydrobiologie, 107.}
\end{thebibliography}
\end{scriptsize}
\end{block}

<<mollusk load data, warning=FALSE>>=
library(PLNmodels); data(mollusk)
mollusk <- 
  prepare_data(mollusk$Abundance, mollusk$Covariate[c("season", "site")]) %>% 
  dplyr::select(-Offset) %>% as_tibble() %>%  distinct() # remove duplicates
mollusk <- cbind(mollusk$Abundance, mollusk[c("season", "site")])
@

\paragraph{External Covariates}

<<mollusc dataset covariates, fig.cap='Natural grouping: site and season'>>=
mollusk %>% dplyr::select(site, season) %>% summary() %>% t() %>% knitr::kable()
@

\end{frame}

\begin{frame}[fragile]
  \frametitle{Companion data set: 'mollusk'}
  \framesubtitle{Brief data summary II}

\paragraph{Histogram of raw counts}

<<mollusc abundance>>=
mollusk %>% dplyr::select(-site, -season) %>% pivot_longer(everything()) %>% 
  ggplot() + aes(x = value, fill = name) + geom_histogram(show.legend = FALSE)
@
\end{frame}

\begin{frame}[fragile]
  \frametitle{Companion data set: 'mollusk'}
  \framesubtitle{Principal Component Analysis}

<<mollusk PCA>>=
mollusk %>% PCA(graph = FALSE, quali.sup = which(map_lgl(mollusk, is.factor))) %>% 
  fviz_pca_biplot(select.var = list(contrib = 5), habillage = "site")
@

\end{frame}

\begin{frame}
  \frametitle{PCA (and linear methods) limitations}

  \begin{block}{Do not account for 'complex' data distribution}
    \begin{itemize}
      \item PCA is tied to a hidden \alert{\bf Gaussian assumption}
      \item Fails with \alert{\bf Count data}
      \item Fails with \alert{\bf Skew data}
    \end{itemize}
  \end{block}
  
  \vfill
  
  \begin{block}{Possible solutions}
    \begin{itemize}
      \item Probabilistic (non Gaussian) models
      \item Need transformed (non-linear) input space
    \end{itemize}
  \end{block}
  
  \end{frame}

\begin{frame}
  \frametitle{Dimension reduction: revisiting the problem setup}

    \begin{block}{Settings}
      \begin{itemize}
        \item \alert{Training data} : $\mathcal{D}=\{\bx_1,\ldots,\bx_n\} \in \Rset^p$,   (i.i.d.)
        \item Space $\Rset^p$ of possibly high dimension $(n \ll p)$
      \end{itemize}
    \end{block}

    \vfill
    
    \begin{block}{Dimension Reduction Map}
       Construct a map $\Phi$ from the space $\Rset^{p}$ into a space $\Rset^{q}$ of \alert{smaller dimension}:
      \begin{align*}
          \Phi:\quad & \Rset^p \to \Rset^{q}, q \ll p\\
                     & \bx \mapsto \Phi(\bx)
      \end{align*}
    \end{block}
    
\end{frame}

\begin{frame}
  \frametitle{How should we design/construct $\Phi$?}

  \paragraph{Geometrical approach} (\alert{\bf see slides on PCA})
  
  \vfill
  
  \paragraph{Idea to go beyond linear approaches}
  \begin{itemize}
    \item Modify the model by amending the \alert{\bf reconstruction error}
    \item Focus on \alert{\bf Relationship preservation}
  \end{itemize}

  \vfill
  
  \paragraph{Form of the map $\Phi$}
  \begin{itemize}
    \item  Linear or \alert{\bf non-linear ?}
    \item tradeoff between  interpretability and \alert{\bf versatility ?}
    \item tradeoff between  \alert{\bf high} or low computational resource
  \end{itemize}

\end{frame}



